test/shared/kaocha/integration_helpers.clj

```diff
@@ -126,6 +129,123 @@
 (defn spit-file [m path contents]
   (let [{:keys [dir] :as m} (test-dir-setup m)
         path (join dir path)]
-    (mkdir (.getParent path))
+    (mkdir (.getParent ^Path path))
     (spit path contents)
     m))
+
+(def ^:dynamic ^Process *process* nil)
+
+(defn interactive-process*
+  [{:keys [dir runner] :as _m} args f]
+  (let [p (-> (doto (ProcessBuilder. ^java.util.List (cons (str runner) args))
+                (.directory (io/file dir)))
+              .start)
+        kill (delay (.destroy p))
+        timeout-ms (cond-> 30000
+                     (System/getenv "CI") (* 2))]
+    (binding [*process* p]
+      (try
+        (let [input (.getOutputStream p)
+              output (.getInputStream p)]
+          (with-open [w (io/writer input)]
+            (with-open [r (io/reader output)]
+              (future
+                ;; unblock read-line calls after 30 seconds and abandon test
+                (Thread/sleep timeout-ms)
+                @kill)
+              (binding [*in* r
+                        *out* w]
+                (f)))))
+        (assert (.waitFor p 10 (java.util.concurrent.TimeUnit/SECONDS))
+                "Process failed to stop!")
+        (.exitValue p)
+        (finally
+          (is (not (realized? kill)) (format "Process was killed after %sms timeout!!" timeout-ms))
```

Comment:
```suggestion
          (is (not (realized? kill)) (format "Process was killed after %sms timeout!" timeout-ms))
```

We usually don't use exclamation points, but if you are going to, use just one. I'm marking all of them if you want to use GitHub's feature to pull in reviewer comments into a commit.